doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+
+##### Error Handling
+
+If any `AsyncHook` callbacks throw, the application will
+print the stack trace and exit. The exit path does follow that of any uncaught
+exception. However `'exit'` callbacks will still fire unless the application
+is run with `--abort-on-uncaught-exception`, in which case a stack trace will
+be printed and the application exits, leaving a core file.
+
+The reason for this error handling behavior is that these callbacks are running
+at potentially volatile points in an object's lifetime. For example, during
+class construction and destruction. Because of this, it is deemed necessary to
+bring down the process quickly in order to prevent an unintentional abort in the
+future. This is subject to change in the future if a comprehensive analysis is
+performed to ensure an exception can follow the normal control flow without
+unintentional side effects.
+
+
+#### `asyncHook.enable()`
+
+* Returns {AsyncHook} A reference to `asyncHook`.
+
+Enable the callbacks for a given `AsyncHook` instance. 
+
+```js
+const async_hooks = require('async_hooks');
+
+const hook = async_hooks.createHook(callbacks).enable();
+```
+
+#### `asyncHook.disable()`
+
+* Returns {AsyncHook} A reference to `asyncHook`.
+
+Disable the callbacks for a given `AsyncHook` instance from the global pool of
+hooks to be executed. Once a hook has been disabled it will not fire again
+until enabled.
+
+For API consistency `disable()` also returns the `AsyncHook` instance.
+
+#### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
+four areas: instantiation, before/after the callback is called, and when the
+instance is destructed. For cases where resources are reused, instantiation and
+destructor calls are emulated.
+
+##### `init(id, type, triggerId, resource)`
+
+* `id` {number} a unique id for the async resource
+* `type` {string} the type of the async resource
+* `triggerId` {number} the unique id of the async resource in whose
+  execution context this async resource was created
+* `resource` {Object} reference to the resource representing the async operation,
+  needs to be released during _destroy_
+
+Called when a class is constructed that has the _possibility_ to trigger an
+asynchronous event. This _does not_ mean the instance must trigger
+`before()`/`after()` before `destroy()` is called. Only that the possibility
+exists.
+
+This behavior can be observed by doing something like opening a resource then
+closing it before the resource can be used. The following snippet demonstrates
+this.
+
+```js
+require('net').createServer().listen(function() { this.close() });
+// OR
+clearTimeout(setTimeout(() => {}, 10));
+```
+
+Every new resource is assigned a unique id.
+
+The `type` is a String that represents the type of resource that caused
+`init()` to fire. Generally it will be the name of the resource's constructor.
+Some examples include `TCP`, `GetAddrInfo` and `HTTPParser`. Users will be able
```

Comment:
Is it too fragile to explicitly enumerate all the types that node core implements? Its not a list that changes very much, and as-is, I think a user of this API would have to experiment with the async_hook API, trying all of the node APIs to see what type of handles they produce. It would be preferable to save the effort, and just document them (again, unless they are changing all the time, or somehow not determinate in the node source - which I don't think is true).